### PR TITLE
[SPARK-30623][Core] Spark external shuffle allow disable of separate event loop group

### PR DESCRIPTION
**Pull request by xuanyuanking**
### What changes were proposed in this pull request?
Fix the regression caused by #22173.
The original PR changes the logic of handling `ChunkFetchReqeust` from async to sync, that's causes the shuffle benchmark regression. This PR fixes the regression back to the async mode by reusing the config `spark.shuffle.server.chunkFetchHandlerThreadsPercent`.
When the user sets the config, ChunkFetchReqeust will be processed in a separate event loop group, otherwise, the code path is exactly the same as before.

As the creation of the separate event loop group is disabled by default, this PR also is a kind of revert for SPARK-25641.

### Why are the changes needed?
Fix the shuffle performance regression described in  https://github.com/apache/spark/pull/22173#issuecomment-572459561

### Does this PR introduce any user-facing change?
Yes, this PR disable the separate event loop for FetchRequest by default.

### How was this patch tested?
Existing UT.

### CODE REVIEW

File: common/network-common/src/main/java/org/apache/spark/network/server/ChunkFetchRequestHandler.java [hunk 1]
```diff
@@ -124,7 +127,13 @@ private ChannelFuture respond(
       final Channel channel,
       final Encodable result) throws InterruptedException {
     final SocketAddress remoteAddress = channel.remoteAddress();
-    return channel.writeAndFlush(result).await().addListener((ChannelFutureListener) future -> {
+    ChannelFuture channelFuture = null;
+    if (chunkFetchAsyncModeEnabled) {
+      channelFuture = channel.writeAndFlush(result);
```
Comment (otterc) [inline common/network-common/src/main/java/org/apache/spark/network/server/ChunkFetchRequestHandler.java]: When we remove `await`, we see more SASL requests timing out. I mentioned this in the comment here:
https://github.com/apache/spark/pull/22173#issuecomment-578347705

Wouldn't making the `asyncMode` default, make this problem worse?
In aysnc mode, how do you plan to tackle increased number of SASL failures?

Comment (xuanyuanking) [reply common/network-common/src/main/java/org/apache/spark/network/server/ChunkFetchRequestHandler.java]: Thanks for the comment @otterc!
This PR aims to fix the performance regression brings by the sync mode fetching. As you mention, you find a problem with the stress test framework, I think the default mode should be the guarantee for common use cases.

For the SASL requests timing out, maybe we need more context for your internal stress test framework and analyze the root cause. IMO, there should be other configs in spark.shuffle/netty could help, not only depends on the async/sync mode here.

**Comment by xuanyuanking** [comment]
cc @cloud-fan @tgravescs

**Comment by tgravescs** [comment]
So I actually filed SPARK-30623 for this, please update the title and things. I don't think we need a separate feature config for this, as that jira mentioned I think was just say if the config isn't explicitly set then do synchronous mode.

**Comment by Victsm** [comment]
@xuanyuanking we worked on the original fix of this issue. Having await in there is the key to the benefits provided in SPARK-24355, which improves reliability of Spark shuffle in a reasonable scaled deployment. This issue seems common across companies like us (LinkedIn), Netflix, Uber, Yahoo. As mentioned in #22173, what we observed is that in cases where HDD is used for shuffle storage, the disk is saturated first before the network can be saturated. So, for a reasonable scaled deployment, having this fix provides a boost in shuffle reliability without hurting much on the performance side. This is also validated by @tgravescs in the Yahoo deployment of this patch.

It's reasonable to introduce another config that disables this reliability improvement if it leads to performance regression in certain deployment mode. Just want to see whether we should leave this enabled by default or not. Also, as mentioned in #22173, we have discovered a potential fix to this perf regression issue that does not removes its reliability benefits. It will take some extra time on our side to evaluate that fix, which is a fix inside Netty. Want to make sure the broader community knows what we have been doing for this issue, so we do not take away a potential reliability improvement to Spark.

**Comment by cloud-fan** [comment]
The default value of `spark.shuffle.server.chunkFetchHandlerThreadsPercent` is 100, so this feature is expected to be disabled by default IIUC.

Policy-wise, we should not fix an issue while introducing a regression. I'm OK to have it since it's disabled by default and it does fix a common issue. What I'm asking for is to make sure the code path is exactly the same as before when this feature is disabled, so that there is no regression.

+1 with @tgravescs to reuse the existing config.

**Comment by xuanyuanking** [comment]
Do some refactoring to reuse the logic of processing fetch requests. Reuse the config `spark.shuffle.server.chunkFetchHandlerThreadsPercent` to control the separate event loop creation. Please take a look.
cc @otterc @Victsm @tgravescs @cloud-fan

File: common/network-common/src/main/java/org/apache/spark/network/util/TransportConf.java [hunk 2]
```diff
@@ -339,12 +341,25 @@ public int chunkFetchHandlerThreads() {
...
     return (int) Math.ceil(threads * (chunkFetchHandlerThreadsPercent / 100.0));
   }
 
+  /**
+   * Whether to use a separate EventLoopGroup to process ChunkFetchRequest messages, it is decided
+   * by the config `spark.shuffle.server.chunkFetchHandlerThreadsPercent` is set or not.
+   */
+  public boolean separateChunkFetchRequest() {
+    try {
+      conf.get("spark.shuffle.server.chunkFetchHandlerThreadsPercent");
```
Comment (cloud-fan) [inline common/network-common/src/main/java/org/apache/spark/network/util/TransportConf.java]: there is no API to check conf existence?

File: common/network-common/src/main/java/org/apache/spark/network/util/TransportConf.java [hunk 3]
```diff
@@ -339,12 +341,25 @@ public int chunkFetchHandlerThreads() {
       return 0;
     }
     int chunkFetchHandlerThreadsPercent =
-      conf.getInt("spark.shuffle.server.chunkFetchHandlerThreadsPercent", 100);
```
Comment (cloud-fan) [inline common/network-common/src/main/java/org/apache/spark/network/util/TransportConf.java]: what's wrong with the previous code?

File: common/network-common/src/main/java/org/apache/spark/network/server/TransportChannelHandler.java [hunk 4]
```diff
@@ -58,19 +58,22 @@
   private final TransportRequestHandler requestHandler;
   private final long requestTimeoutNs;
   private final boolean closeIdleConnections;
+  private final boolean separateChunkFetchRequest;
```
Comment (cloud-fan) [inline common/network-common/src/main/java/org/apache/spark/network/server/TransportChannelHandler.java]: maybe a more explicit name: `skipChunkFetchRequest`

File: common/network-common/src/main/java/org/apache/spark/network/server/ChunkFetchRequestHandler.java [hunk 5]
```diff
@@ -17,46 +17,33 @@
...
+ * See more details in SPARK-30623.
  */
-public class ChunkFetchRequestHandler extends SimpleChannelInboundHandler<ChunkFetchRequest> {
+public abstract class ChunkFetchRequestHandler
+    extends SimpleChannelInboundHandler<ChunkFetchRequest> {
   private static final Logger logger = LoggerFactory.getLogger(ChunkFetchRequestHandler.class);
 
   private final TransportClient client;
   private final StreamManager streamManager;
-  /** The max number of chunks being transferred and not finished yet. */
```
Comment (cloud-fan) [inline common/network-common/src/main/java/org/apache/spark/network/server/ChunkFetchRequestHandler.java]: shall we keep the comment?

File: common/network-common/src/main/java/org/apache/spark/network/server/TransportRequestHandler.java [hunk 6]
```diff
@@ -67,6 +68,7 @@ public TransportRequestHandler(
       TransportClient reverseClient,
       RpcHandler rpcHandler,
       Long maxChunksBeingTransferred) {
+    super(reverseClient, rpcHandler.getStreamManager(), maxChunksBeingTransferred);
```
Comment (cloud-fan) [inline common/network-common/src/main/java/org/apache/spark/network/server/TransportRequestHandler.java]: do we still need the `maxChunksBeingTransferred` variable in this class?

File: common/network-common/src/main/java/org/apache/spark/network/server/ChunkFetchRequestHandler.java [hunk 7]
```diff
@@ -107,31 +85,10 @@ protected void channelRead0(
...
-      if (future.isSuccess()) {
-        logger.trace("Sent result {} to client {}", result, remoteAddress);
-      } else {
-        logger.error(String.format("Error sending result %s to %s; closing connection",
-          result, remoteAddress), future.cause());
-        channel.close();
-      }
-    });
-  }
+  public abstract ChannelFuture respond(Channel channel, Encodable result) throws Exception;
```
Comment (cloud-fan) [inline common/network-common/src/main/java/org/apache/spark/network/server/ChunkFetchRequestHandler.java]: why we need to define this method?

File: common/network-common/src/main/java/org/apache/spark/network/server/ChunkFetchRequestHandler.java [hunk 8]
```diff
@@ -65,17 +52,8 @@ public ChunkFetchRequestHandler(
...
-    ctx.close();
-  }
-
-  @Override
-  protected void channelRead0(
-      ChannelHandlerContext ctx,
-      final ChunkFetchRequest msg) throws Exception {
-    Channel channel = ctx.channel();
+  public void processFetchRequest(
+      final Channel channel, final ChunkFetchRequest msg) throws Exception {
```
Comment (cloud-fan) [inline common/network-common/src/main/java/org/apache/spark/network/server/ChunkFetchRequestHandler.java]: shall we define channel as a protected member variable?

File: common/network-common/src/main/java/org/apache/spark/network/util/TransportConf.java (hunk 2 above)
Comment (xuanyuanking) [reply common/network-common/src/main/java/org/apache/spark/network/util/TransportConf.java]: Yes, no API in `ConfigProvider`.

File: common/network-common/src/main/java/org/apache/spark/network/util/TransportConf.java (hunk 3 above)
Comment (xuanyuanking) [reply common/network-common/src/main/java/org/apache/spark/network/util/TransportConf.java]: No need to give a default value here, when it comes to here, the config must be set.

File: common/network-common/src/main/java/org/apache/spark/network/server/TransportChannelHandler.java (hunk 4 above)
Comment (xuanyuanking) [reply common/network-common/src/main/java/org/apache/spark/network/server/TransportChannelHandler.java]: Thanks

File: common/network-common/src/main/java/org/apache/spark/network/server/ChunkFetchRequestHandler.java (hunk 5 above)
Comment (xuanyuanking) [reply common/network-common/src/main/java/org/apache/spark/network/server/ChunkFetchRequestHandler.java]: Sure.

File: common/network-common/src/main/java/org/apache/spark/network/server/ChunkFetchRequestHandler.java (hunk 8 above)
Comment (xuanyuanking) [reply common/network-common/src/main/java/org/apache/spark/network/server/ChunkFetchRequestHandler.java]: It's OK for `TransportRequestHandler`, but for `ChunkFetchChannelHandler`, maybe get the current channel from `ChannelHandlerContext` is a safer way.

File: common/network-common/src/main/java/org/apache/spark/network/server/TransportRequestHandler.java (hunk 6 above)
Comment (xuanyuanking) [reply common/network-common/src/main/java/org/apache/spark/network/server/TransportRequestHandler.java]: Yes, it is still in use in `processStreamRequest`.

File: common/network-common/src/main/java/org/apache/spark/network/util/TransportConf.java (hunk 2 above)
Comment (cloud-fan) [reply common/network-common/src/main/java/org/apache/spark/network/util/TransportConf.java]: how about `conf.getInt("spark.shuffle.server.chunkFetchHandlerThreadsPercent", 0) > 0`?

File: common/network-common/src/test/java/org/apache/spark/network/TransportRequestHandlerSuite.java [hunk 9]
```diff
@@ -39,7 +39,7 @@
 public class TransportRequestHandlerSuite {
 
   @Test
-  public void handleStreamRequest() {
+  public void handleStreamRequest() throws Exception {
```
Comment (cloud-fan) [inline common/network-common/src/test/java/org/apache/spark/network/TransportRequestHandlerSuite.java]: is this change necessary?

Comment (xuanyuanking) [reply common/network-common/src/test/java/org/apache/spark/network/TransportRequestHandlerSuite.java]: Yes, because of the changes for `TransportRequestHandler.handler` here: https://github.com/apache/spark/pull/27665/files#diff-0e3429029d3f8d49e94ef11e4e3051a2R105

**Comment by cloud-fan** [comment]
retest this please

File: common/network-common/src/main/java/org/apache/spark/network/util/TransportConf.java (hunk 2 above)
Comment (xuanyuanking) [reply common/network-common/src/main/java/org/apache/spark/network/util/TransportConf.java]: Thanks, done in 4f42083.

**Review by cloud-fan — APPROVED** [review]

**Comment by xuanyuanking** [comment]
retest this please

**Comment by cloud-fan** [comment]
thanks, merging to master/3.0!

**Comment by dongjoon-hyun** [comment]
Hi, @cloud-fan . This seems to be not in `branch-3.0` yet.

**Comment by cloud-fan** [comment]
it is now. merge script runs slowly at my sides...

**Comment by dongjoon-hyun** [comment]
Thank you!

File: common/network-common/src/main/java/org/apache/spark/network/util/TransportConf.java (hunk 3 above)
Comment (dongjoon-hyun) [reply common/network-common/src/main/java/org/apache/spark/network/util/TransportConf.java]: What do you mean by `the config must be set`, @xuanyuanking ? What value do you expect by default? Apparently, this is reverting SPARK-25641 without mentioning SPARK-25641. In the PR, only SPARK-24355 is mentioned.
> No need to give a default value here, when it comes to here, the config must be set.

**Comment by dongjoon-hyun** [comment]
BTW, @xuanyuanking .
Could you confirm the above question?
- https://github.com/apache/spark/pull/27665#discussion_r398328411

File: common/network-common/src/main/java/org/apache/spark/network/util/TransportConf.java (hunk 3 above)
Comment (cloud-fan) [reply common/network-common/src/main/java/org/apache/spark/network/util/TransportConf.java]: because we only call this method if `separateChunkFetchRequest` returns true.

We will see exception if the assumption is broken.

Comment (xuanyuanking) [reply common/network-common/src/main/java/org/apache/spark/network/util/TransportConf.java]: In this PR, we make the separate event loop group configurable by checking the config `spark.shuffle.server.chunkFetchHandlerThreadsPercent` is set or not.
> What do you mean by the config must be set

Here the function `chunkFetchHandlerThreads` is only called while the config is set. https://github.com/apache/spark/blob/0fe203e7032a326ff0c78f6660ab1b09409fe09d/common/network-common/src/main/java/org/apache/spark/network/TransportContext.java#L124-L129

> What value do you expect by default? Apparently, this seems to revert SPARK-25641 together without mentioning SPARK-25641. 

Yes, this PR makes the feature disabled by default, let me also mention SPARK-25641 in PR description.

**Comment by xuanyuanking** [comment]
Thanks for the review.

File: common/network-common/src/main/java/org/apache/spark/network/util/TransportConf.java (hunk 3 above)
Comment (dongjoon-hyun) [reply common/network-common/src/main/java/org/apache/spark/network/util/TransportConf.java]: Thank you, @xuanyuanking and @cloud-fan .
